### PR TITLE
fix(executor/emissary): fix nonroot sidecars + input/output params & artifacts

### PR DESCRIPTION
**Pull request by Bobgy**
This fully fixes emissary + nonroot.
I've tested sidecars, input/output parameters and artifacts

Checklist:

* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md).

Tips:

* Your PR needs to pass the required checks before it can be approved. If the check is not required (e.g. E2E tests) it does not need to pass
* Sign-off your commits to pass the DCO check: `git commit --signoff`.
* Run `make pre-commit -B` to fix codegen or lint problems. 
* Say how how you tested your changes. If you changed the UI, attach screenshots.

### CODE REVIEW

File: cmd/argoexec/commands/emissary.go [hunk 1]
```diff
@@ -39,14 +39,21 @@ func NewEmissaryCommand() *cobra.Command {
...
 				if err != nil {
 					logger.Error(fmt.Errorf("failed to write exit code: %w", err))
 				}
 			}()
 
-			// this also indicates we've started
-			if err := os.MkdirAll(varRunArgo+"/ctr/"+containerName, 0o700); err != nil {
+			// default umask can be 022
+			// setting umask as 0 allow granting write access to other non-root users
+			syscall.Umask(0)
```
Comment (Bobgy) [inline cmd/argoexec/commands/emissary.go:50]: Should I put this into os_specific?

**Comment by Bobgy** [comment]
Hi @alexec, I wonder whether you have any docs or examples for how to add workflows as tests?
I think it'll be a lot more confident checking in workflows to verify the fix and make sure we don't break it in the future.

**Comment by Bobgy** [comment]
nonroot sidecar example:
```
#### sidecar example ####
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: nonroot-sidecar-
spec:
  serviceAccountName: pipeline-runner
  entrypoint: sidecar-example
  templates:
  - name: sidecar-example
    container:
      image: alpine:latest
      command: [sh, -c]
      args: ["
        apk update &&
        apk add curl &&
        until curl 'http://127.0.0.1:8080'; do sleep 5; done &&
      "]
    sidecars:
    - name: http-echo
      image: mendhak/http-https-echo:19
      command: ['docker-entrypoint.sh']
      args: ['node', './index.js']
      envs:
      - name: HTTP_PORT
        value: 8080
```

nonroot parameter passing
```
#### parameter passing test workflow for non-root ####
# Output parameters provide a way to use the contents of a file,
# as a parameter value in a workflow. In that regard, they are
# similar in concept to script templates, with the difference being
# that the output parameter values are obtained via file contents
# instead of stdout (as with script templates). Secondly, there can
# be multiple 'output.parameters.xxx' in a single template, versus
# a single 'output.result' from a script template.
# 
# In this example, the 'whalesay' template produces an output
# parameter named 'hello-param', taken from the file contents of
# /tmp/hello_world.txt. This parameter is passed to a subsequent
# step as an input parameter to the template, 'print-message'.
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: nonroot-parameter-
spec:
  serviceAccountName: pipeline-runner
  entrypoint: output-parameter
  templates:
  - name: output-parameter
    steps:
    - - name: generate-parameter
        template: whalesay
    - - name: consume-parameter
        template: print-message
        arguments:
          parameters:
          - name: message
            value: "{{steps.generate-parameter.outputs.parameters.hello-param}}"

  - name: whalesay
    container:
      image: curlimages/curl:latest
      command: [sh, -c]
      args: ["sleep 1; echo -n hello world > /tmp/hello_world.txt"]
    outputs:
      parameters:
      - name: hello-param
        valueFrom:
          default: "Foobar"   # Default value to use if retrieving valueFrom fails. If not provided workflow will fail instead
          path: /tmp/hello_world.txt

  - name: print-message
    inputs:
      parameters:
      - name: message
    container:
      image: curlimages/curl:latest
      command: [sh, -c]
      args: ["echo {{inputs.parameters.message}}"]
```

non root hello world workflow:
```
#### one step test workflow for non-root ####
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: nonroot-helloworld-
spec:
  serviceAccountName: pipeline-runner
  entrypoint: hw
  templates:
  - name: hw
    container:
      image: curlimages/curl:latest
      command: [sh, -c, "echo hello"]
```

nonroot artifact passing workflow:
```
#### artifact passing test workflow for non-root ####
# This example demonstrates the ability to pass artifacts
# from one step to the next.

apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: nonroot-artifact-
spec:
  serviceAccountName: pipeline-runner
  entrypoint: artifact-example
  templates:
  - name: artifact-example
    steps:
    - - name: generate-artifact
        template: whalesay
    - - name: consume-artifact
        template: print-message
        arguments:
          artifacts:
          - name: message
            from: "{{steps.generate-artifact.outputs.artifacts.hello-art}}"

  - name: whalesay
    container:
      image: curlimages/curl:latest
      command: [sh, -c]
      args: ["sleep 1; echo hello world | tee /tmp/hello_world.txt"]
    outputs:
      artifacts:
      - name: hello-art
        path: /tmp/hello_world.txt

  - name: print-message
    inputs:
      artifacts:
      - name: message
        path: /tmp/message
    container:
      image: curlimages/curl:latest
      command: [sh, -c]
      args: ["cat /tmp/message"]
```

File: cmd/argoexec/commands/emissary.go (hunk 1 above)
Comment (alexec) [reply cmd/argoexec/commands/emissary.go:50]: `syscall.Umask(0)` is not available on Windows. So you'll need to move this command into `os_specific`. The Windows implementation can be a no-op.

**Review by alexec — CHANGES_REQUESTED** [review]
great stuff! you're correct about Umask - this needs to go into os_specific

**Comment by Bobgy** [comment]
@alexec thanks for the review! Updated to os specific

**Comment by Bobgy** [comment]
Any guidance for
> Hi @alexec, I wonder whether you have any docs or examples for how to add workflows as tests?
I think it'll be a lot more confident checking in workflows to verify the fix and make sure we don't break it in the future.
?

File: workflow/executor/os-specific/umask_darwin.go [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package os_specific
+
+import "syscall"
+
+func CallUmask(mask int) (oldmask int) {
```
Comment (alexec) [inline workflow/executor/os-specific/umask_darwin.go:5]: minor - you ignore the return value and you only ever pass in zero - so how about:

```
func UmaskZero() {
// ...
}
```

What might be even better is name it after you intent.

Comment (alexec) [reply workflow/executor/os-specific/umask_darwin.go:5]:
```
GrantWriteAccessToNonRootUsers
```

??

Comment (Bobgy) [reply workflow/executor/os-specific/umask_darwin.go:5]: What do you think about AllowGrantingAccessToEveryone?

Comment (Bobgy) [reply workflow/executor/os-specific/umask_darwin.go]: OK, updated

**Review by alexec — APPROVED** [review]

**Comment by alexec** [comment]
@sarabala1979 this needs to be back-ported to v3.1. Do you think we should do another patch release this week?

**Pull request merged**